### DISCUSSION TITLE
How to set global variables like `COMPOSE_FILE` and `IMAGE_NAME` across reusable workflows and custom actions?

### DISCUSSION THREAD

**Discussion by dunk** [community/community#30937] [Actions]
In a common CI pattern the following env vars are required by `docker compose`

```
COMPOSE_FILE=docker-compose.yml:docker-compose.ci.yml
IMAGE_NAME=ghcr.io/aaa/bbb/ccc:pull_request_identifier
```

In this supposed setup the docker-compose.ci.yml would define:

```
services:
  someservice:
    image: ${IMAGE_NAME}
```

So `COMPOSE_FILE` and `IMAGE_NAME` are required for all interactions with `docker compose`.

If we want to separate out some parts of our workflow for reuse, this becomes very tedious and copy-paste-y:

* Every top-level workflow must define `COMPOSE_FILE` and `IMAGE_NAME`, which is duplicative.
* Every time we create a custom action that will be called by more than one step, we have to pass both variables every time. And, tediously, we need to run a step to re-echo them out into the `$GITHUB_ENV` for that step.

A nice little hidden gotcha here is that if you try to use a reusable workflow instead of a custom action, you can't actually pass env vars to it at all, as it's being run in the wrong phase and env vars aren't available.

Am I missing something here? It feels like there should be a way to set some kind of global variable that will work across whatever reusable workflows or custom steps that you decide to import? And, separately, is there some standard way to set things like `COMPOSE_FILE` across workflows (in a way that's not a hack like schlepping them out of a random file and appending to `$GITHUB_ENV`?

**Comment by laughedelic** [comment]
AFAIK, there is no way to set global constants for GitHub actions. But you can use repo or org _secrets_ for that. It's not transparent, because you won't be able to see the secret value once you create it. It might be also tricky to use in shared actions and reusable workflows, because you'd need to add them through the inputs. So it's more of a workaround than a solution.

**Reply by dunk** [reply]
That's troubling. There's presumably millions of repos with this setup... In fact, ignoring trivial cases, isn't this like the most common case? That or very close to it. Does that mean everyone is using clunky, repetitive, fragile workarounds too then?
